devenv/main.py

```diff
@@ -67,47 +33,48 @@ def devenv(argv: Sequence[str], config_path: str) -> ExitCode:
     config = read_config(config_path)
 
     # Guessing temporary code root
-    coderoot = config.get("devenv", "coderoot", fallback=None) or (
+    code_root = config.get("devenv", "coderoot", fallback=None) or (
         os.path.abspath(f"{current_root}/..")
         if current_root
         else os.path.expanduser("~/code")
     )
 
-    if args.command == "bootstrap":
-        return bootstrap.main(config_path, coderoot, remainder)
+    modinfo_list = [
+        (module.__name__, module.module_info)
+        for module in [bootstrap, fetch, doctor, pin_gha, sync]
+        if hasattr(module, "module_info")
+    ]
 
-    if args.command == "fetch":
-        return fetch.main(coderoot, remainder)
+    parser = argparse.ArgumentParser()
+    subparser = parser.add_subparsers(
+        title=argparse.SUPPRESS, metavar="command", dest="module", required=True
+    )
 
-    if not args.nocoderoot and not os.getcwd().startswith(coderoot):
-        print(
-            f"You aren't in your code root ({coderoot})!\n"
-            "To ignore, use devenv --nocoderoot [COMMAND]\n"
-            f"To change your code root, you can edit {config_path}.\n"
+    for name, info in modinfo_list:
+        # Argparse stuff
+        child = subparser.add_parser(info.command, help=info.help)
+        child.add_argument(
+            "--xixax",
+            dest="_module_exec",
+            default=info.action,
+            required=False,
+            help=argparse.SUPPRESS,
         )
-        return 1
 
-    # the remaining tools are repo-specific
-    reporoot = gitroot()
+    args, remainder = parser.parse_known_args(argv[1:])
 
-    fake_reporoot = os.getenv("CI_DEVENV_INTEGRATION_FAKE_REPOROOT")
-    if fake_reporoot:
-        reporoot = fake_reporoot
-
-    repo = reporoot.split("/")[-1]
-    context = {"repo": repo, "reporoot": reporoot}
+    # context for subcommands
+    context: Context = {
+        "config_path": config_path,
+        "code_root": code_root,
+        "repo": Repository(current_root) if current_root else None,
+    }
 
-    if args.command == "doctor":
-        return doctor.main(context, remainder)
-    if args.command == "sync":
-        return sync.main(context, remainder)
-
-    return 1
+    return cast(ExitCode, args._module_exec(context, remainder))
```

Comment:
`cast` has runtime overhead and generally should only be used for bugs in the type checker -- https://www.notion.so/sentry/sentry-typing-guidelines-best-practices-471d5aacf262476188982ce70f909e00?pvs=4#4c5f66d5e5f2460991cf1074111fdd85

Reply:
I think I picked this up from sync(). I'll remove it and check to make sure it still works.